cmd/complex/logger.go

```diff
@@ -2,77 +2,46 @@ package main
 
 import (
 	"context"
-	"fmt"
-	"log"
+	"errors"
+	"log/slog"
 	"net/http"
-	"strings"
-
-	"github.com/tmaxmax/go-sse"
+	"os"
 )
 
-type logger struct {
-	l *log.Logger
-}
-
-// Log is a logger for sse.Server. It logs events like so:
-//
-//	level=INFO msg="A message about the logged event"
-//	  ua="the user agent" ip="the RemoteAddr" host="the host" origin="the origin"
-//	  lastEventID=5 topics=""
-//
-// It expects that the given context contains a value mapped to loggerCtxKey{},
-// which will happen because sse.Server passes the request context and we set
-// a middleware up to add these to the context.
-func (l logger) Log(ctx context.Context, level sse.LogLevel, msg string, data map[string]any) {
-	levelStr := ""
-	switch level {
-	case sse.LogLevelInfo:
-		levelStr = "INFO"
-	case sse.LogLevelWarn:
-		levelStr = "WARN"
-	case sse.LogLevelError:
-		levelStr = "ERROR"
-	}
-
-	output := fmt.Sprintf("level=%s msg=%q", levelStr, msg)
-	if e, ok := data["err"]; ok {
-		output += fmt.Sprintf(" err=%q", e.(error).Error())
-	}
-
-	delete(data, "err")
-
-	output += "\n  "
-
-	for k, v := range ctx.Value(loggerCtxKey{}).(map[string]string) {
-		output += fmt.Sprintf(" %s=%q", k, v)
-	}
-
-	output += "\n  "
-
-	for k, v := range data {
-		switch s := v.(type) {
-		case sse.EventID: // key is "lastEventID"
-			output += fmt.Sprintf(" %s=%q", k, s)
-		case []string: // key is "topics"
-			output += fmt.Sprintf(" %s=%q", k, strings.Join(s, ", "))
-		default:
-			panic(fmt.Errorf("unexpected value type %T", v))
-		}
-	}
-
-	l.l.Println(output)
-}
-
 type loggerCtxKey struct{}
 
+// withLogger is an http middleware that generates a logger with request-specific fields
+// added to it and attaches it to the request context for later retrieval with getLogger().
+// This is simmilar to how existing packages like https://github.com/go-chi/httplog works.
```

Comment:
I'd remove links to external packages. They look like endorsements to me and I don't think that's appropriate. Comments should also use correct initialism for acronyms ("HTTP", not "http").
```suggestion
// withLogger is an HTTP middleware that creates a logger with request-specific fields
// added to it and attaches it to the request context for later retrieval with getLogger().
```

Reply:
This is one of those "specific beats general" cases in my mind, since it's the HTTP protocol but I'm referencing the more-or-less standard pattern for middleware for the stdlib `http` package.

I did remove the specific reference for a more general reference to 3rd party libraries though.

Reply:
Oh, I see what you mean – maybe in that case we could write "net/http" or "stdlib-compatible" instead of just "http"? The latter is too easy to brush off as just the HTTP protocol, as I initially did.

Pertaining to the 3rd party references, I'd probably be a bit more specific on how they're similar and rephrase the comments to:
- for `withLogger`: "Third party logging packages may offer similar middlewares to add a logger to the request or maybe just a helper to add a logger to context; in the second case you can build your own middleware function around it, similar to this one."
- for `getLogger`: "Third party logging packages may offer similar helpers for retrieving a logger from a request or a context."

Reply:
✅